man/gricli.1
.TH GRICLI 1 "July 1, 2011" "Gricli user manual"
.SH NAME
.PP
gricli - grid commandline interface
.SH SYNOPSIS
.PP
gricli [\f[I]options\f[]]
.SH DESCRIPTION
.PP
Gricli is a shell that allows commandline interaction with the
grid.
You can use it to submit, control and monitor jobs.
It also supports easy access to grid-filesystems and resource
information.
.PP
Gricli is based on the \f[I]Grisu\f[] framework and can connect to
different \f[I]Grisu backends\f[] by selecting the appropriate one
as a commandline parameter, e.
g.:
.PP
\f[CR]
      gricli\ -b\ BeSTGRID
\f[]
.PP
or
.PP
\f[CR]
      gricli\ -b\ Local
\f[]
.PP
The above first would connect to the default \f[I]BeSTGRID\f[]
backend that publishes the \f[I]Grisu\f[] API via SOAP.
The latter would connect to a local \f[I]Grisu\f[] backend which
sits on the same computer as \f[I]gricli\f[], as long as the local
backend jar
(http://code.ceres.auckland.ac.nz/downloads/local-backend.jar) is
in the classpath (either in the same folder as gricli or in
%$HOME/.grisu.beta/lib/).
.SH OPTIONS
.PP
-b \f[I]BACKEND\f[] or --backend=\f[I]BACKEND\f[]
.PP
The Grisu backend to connect to.
The default is \f[I]BeSTGRID\f[] and other possible backends are
\f[I]BeSTGRID-TEST\f[],\f[I]BeSTGRID-DEV\f[] and \f[I]Local\f[].
.PP
Examples:
.PP
\f[CR]
      gricli\ -b\ BeSTGRID
      gricli\ --backend=BeSTGRID
\f[]
.PP
-f \f[I]SCRIPT\f[] or --file=\f[I]SCRIPT\f[]
.PP
Executes a gricli script.
.PP
Examples:
.PP
\f[CR]
      gricli\ -f\ myexp.gs
      gricli\ --file=myexp.gs
\f[]
.PP
-n or --nologin
.PP
Disables login at gricli startup.
.PP
Example:
.PP
\f[CR]
      gricli\ -n
\f[]
.SH GLOBALS
.PP
Globals define the properties of a job.
.SS Overview
.PP
Globals are the properties for a job.
They let you determine how and where your job will be executed.
To see the list of globals use the command `print globals'.
You may use the `set' command to set any of the global properties.
To view an individual property use the command `print global '.
.PP
To view information and examples on each global use the command
`help '.
Note that you may use a plain text script to automate the task of
setting globals.
For more information see use the command `help run'.
.PP
Example usage:
.PP
print globals print global memory set memory 1g help memory help
global memory
.SS List of globals:
.SS cpus
.PP
The number of CPUs to be used by a job.
.PP
To set the number of CPUs use the `set' command.
To view the number of CPUs used in a job use the command
`print global cpus'.
After a job has been submitted you can check the CPUs used with the
command `print job  cpus'
.PP
Example usage:
.PP
\f[CR]
      set\ cpus\ 10
      print\ global\ cpus
      print\ job\ myjob\ cpus
\f[]
.SS debug
.PP
Boolean to show debug output from exceptions.
.PP
This value can be set using the `set' command.
.PP
Example usage:
.PP
\f[CR]
      set\ debug\ true
      set\ debug\ false
\f[]
.SS description
.PP
The description of a job.
.PP
To set the job description use the `set' command.
To view the description before the job has been submitted use the
command `print global description'.
To view the description of a submitted job use the command
`print job  description'.
.PP
Example usage:
.PP
\f[CR]
      set\ description\ "my\ job\ description"
      print\ global\ description
      print\ job\ myjob\ description
\f[]
.SS dir
.PP
The job directory.
.PP
This is the directory the job will be downloaded to after a
`download' or `downloadclean' command if a target directory is not
specified.
.PP
It is also used as the starting point where relative paths are
applicable.
For example if a file is located at /home/myfolder/myfile and the
dir variable is /home then myfile can be attached with using the
relative path: attach myfolder/myfile
.PP
To set the dir use the `set' command or the `cd' command.
To view the dir before a job has been submitted use the command
`print global dir'.
To view the dir after a job as been submitted use the command
`print job  jobDirectory'.
.PP
Grid locations (starting with prefix grid://) are currently not
supported for this command.
.PP
Example usage:
.PP
\f[CR]
      set\ dir\ ~
      set\ dir\ /home/myfolder
      cd\ ~/myfolder
\f[]
.SS email
.PP
The email address to send notifications to.
.PP
Email notifications can be sent when a job has started and when it
has finished.
.PP
The email address can be set using the `set' command.
To view the email address of a job before submission use the
command `print global email'.
To view the email address of a job after submission use the command
`print job  email_address'.
.PP
Example usage:
.PP
\f[CR]
      set\ email\ myemail\@myhost.x
      print\ global\ email
\f[]
.SS email_on_finish
.PP
Boolean to send an email notification when a job has finished.
.PP
To set the variable use the `set' command.
To view the setting before submission use the command
`print global email_on_finish'.
To view setting after submission use the command
`print job  email_on_finish'.
.PP
Example usage:
.PP
\f[CR]
      set\ email_on_finish\ true
      set\ email_on_finish\ false
\f[]
.SS email_on_start
.PP
Boolean to send an email notification when a job has started
executing.
.PP
To set the variable use the `set' command.
To view the setting before submission use the command
`print global email_on_start'.
To view setting after submission use the command
`print job  email_on_start'.
.PP
Example usage:
.PP
\f[CR]
      set\ email_on_start\ true
      set\ email_on_start\ false
\f[]
.SS env
.PP
The execution environment variables of a job.
.PP
To add an environment variable and value use the `add env  '
command.
.PP
Note that you do not need `$' as part of the variable name.
.PP
To view the environment variables and their values before
submission use the command `print global env'.
To view the environment variables after submission use the command
`print job  env'.
.PP
Example usage:
.PP
\f[CR]
      add\ env\ MY_VAR\ MY_VALUE
      print\ global\ env
      print\ job\ myjob\ env
\f[]
.PP
For MPI jobs using multiple hosts, the environment variables must
be explicitly exported using the -x option in mpirun e.g:
.PP
\f[CR]
      submit\ -x\ MY_VAR\ /home/me001/my_application\ arg0\ arg1
\f[]
.SS gdir
.PP
The grid directory.
.PP
This is for use with the filemanager command which has yet to be
implemented.
.SS group
.PP
The group used to send jobs.
.PP
The group determines which queues you will have access to and
consequently which application package you can use.
.PP
To set the group use the `set' command.
Note that the group must be set before a job can be submitted.
To view the group before a job has been submitted use the command
`print global group'.
To view the group after a job has been submitted use the command
`print job  group'.
.PP
Example usage:
.PP
\f[CR]
      set\ group\ /nz/nesi
      print\ global\ group
      print\ job\ myjob\ group
\f[]
.SS hostcount
.PP
The number of compute hosts to be used
.PP
The hostcount is important for jobs where processes communicate
across a number of physical machines or hosts e.g.\ MPI jobs.
Setting the hostcount will force the job to use the set number of
hosts.
This can improve efficiency as the communications overhead is less
between processes running on the same host.
However the job may take longer to be dequeued as the requirements
are more restrictive.
.PP
The hostcount is unset by default and will not show in the list of
globals.
Once set, it will be visible in the list of globals.
Note that when setting the hostcount, you must use a positive
integer.
To disable the hostcount restriction use the command
`unset hostcount'.
.PP
Example usage:
.PP
\f[CR]
      set\ hostcount\ 2
      unset\ hostcount
      print\ global\ hostcount
      print\ job\ myjob\ hostcount
\f[]
.SS jobname
.PP
The job name.
.PP
This will be the name of the submitted job.
If a job with this name already exists, an integer will be appended
to make sure it is unique.
.PP
To set the job name use the `set' command.
To view the job name before submission use the command
`print global jobname'.
To view the job name after a job has been submitted use the command
`print jobs'.
.PP
Example usage:
.PP
\f[CR]
      set\ jobname\ myjob
      print\ global\ jobname
\f[]
.SS jobtype
.PP
The type of job to submit.
.PP
The job type determines how the job is configured for execution.
.PP
The current values are:
.PP
\f[CR]
      single\ :\ A\ job\ that\ will\ use\ one\ CPU\ on\ one\ host.
      smp\ \ \ \ :\ A\ job\ that\ will\ use\ one\ or\ more\ CPUs\ on\ one\ host.
      mpi\ \ \ \ :\ A\ job\ that\ will\ use\ one\ or\ more\ CPUs\ across\ one\ or\ more\ hosts\ using\ the\ Open\ MPI\ framework.
\f[]
.PP
Please note that a `host' is a compute node within a queue.
Since the hardware specifications may vary between hosts in a
queue, you are advised to check the properties of the queue to
ensure you jobs run correctly.
In particular, it is important that jobs do not request more
resources than are available for a given job type.
.PP
By default, an mpi job may schedule CPUs on any nodes in the queue.
You may use the hostcount global to force the CPUs to be scheduled
on a specific number of nodes.
To remove this restriction, use the unset command:
.PP
\f[CR]
      set\ hostcount\ 2
      unset\ hostcount
\f[]
.PP
If you have set the hostcount, you can check the value using the
command `print global hostcount' and after submission using the
command `print job  hostcount'.
.PP
Example usage:
.PP
\f[CR]
      set\ jobtype\ mpi
      print\ global\ jobtype
      print\ job\ myjob\ hostcount
\f[]
.SS memory
.PP
The total memory (in MB) to be used by the job.
.PP
The value of this global represents the amount of physical memory
(RAM) to be allocated as well as the amount of virtual memory to be
allocated.
This means that if you enter the following command:
.PP
\f[CR]
      set\ memory\ 1024
\f[]
.PP
Your job will have 1024 MB (or 1 GB) of RAM and 1 GB of virtual
memory
.PP
The way memory is used depends on the jobtype.
.PP
\f[CR]
      single\ :\ All\ memory\ is\ used\ by\ one\ CPU.
      smp\ \ \ \ :\ The\ memory\ is\ shared\ between\ one\ or\ more\ CPUs\ on\ a\ single\ host.
      mpi\ \ \ \ :\ The\ memory\ is\ divided\ between\ the\ CPUs\ which\ may\ be\ on\ one\ or\ more\ hosts.
\f[]
.PP
To set the memory for the job, use the `set' command.
The command accepts values in the following formats:
.PP
\f[CR]
      set\ memory\ 200\ \ \ \ :\ sets\ memory\ to\ 200\ MB
      set\ memory\ 200m\ \ \ :\ sets\ memory\ to\ 200\ MB
      set\ memory\ 1g\ \ \ \ \ :\ sets\ memory\ to\ 1\ GB\ (1024\ MB)
      set\ memory\ 1g200m\ :\ sets\ memory\ to\ 1224\ MB
\f[]
.PP
To view the memory of a job before submission use the command
`print global memory'.
To view the memory of a job after submission use the command 'print
job memory.
.PP
Please note that if you request more memory than is available for
your jobtype on a given queue, the job may stay on the queue
because the scheduler cannot find the appropriate resources to
start the job.
.PP
Example usage:
.PP
set memory 1224 set memory 1g200m print global memory print job
myjob memory
.SS outputfile
.PP
The path to a file where command output is redirected to.
.PP
Some commands will print messages for the user.
This output can be redirected to a file for processing.
.PP
Note that this option does not redirect job output.
They will use the standard output files stdout.txt and stderr.txt
You can see the contents of these files using the `view' command
e.g:
.PP
\f[CR]
      view\ myjob\ stdout.txt
      view\ myjob\ stderr.txt
\f[]
.PP
Example usage:
.PP
\f[CR]
      set\ outputfile\ /home/myfolder/output.txt
\f[]
.SS package
.PP
This is the application package used by the job.
.PP
To set the package use the `set' command.
To see a list of available packages use the `print packages'
command.
.PP
To see which package is set for a job before it is submitted, use
the command `print global package'.
After a job has been submitted you can check the package with
`print job  package'
.PP
Note that the package is set to generic by default.
If you want the queue to be determined automatically, then it is
best to set the package to ensure that the selected queue can
support your job.
Otherwise you would need to set the queue manually and check that
it supports the application run by your job.
If you would like to set the queue manually, use the command
`print package ' to see the available queues for your application.
.PP
Example usage:
.PP
\f[CR]
      print\ package
      set\ package\ R
      print\ global\ package
      print\ job\ myjob\ package
\f[]
.SS prompt
.PP
The prompt message.
.PP
This is can changed using the `set' command.
The command can accept macros, substituting the values of other
globals into the prompt.
.PP
Example usage:
.PP
set prompt \[lq]myprompt>\[rq] set prompt
\[lq]\f[B]{dir}>\ "\ set\ prompt\ "\f[]{jobname}>\[rq]
.SS queue
.PP
The job queue.
.PP
The job queue will determine which resources and application
packages are available for your job.
.PP
To set the queue use the `set' command.
To see a list of queues use the `print queues command'.
.PP
You can only submit jobs to queues assigned to your group.
To view the available groups use the `print groups' command.
To view the queues available for a specific groups use the command
`print queues '
.PP
To see which queues support a particular application package use
the command `print package <application_package>'.
To see a list of application packages use the command
`print packages'.
.PP
If you have set the application package, then the queue location
can be determined automatically.
Use the command `set queue auto' to enable this option.
.PP
To see the queue before a job is submitted use the command
`print global queue'.
To see the queue after a job has been submitted use the command
`print job  submissionLocation'.
.PP
Example usage:
.PP
\f[CR]
      set\ queue\ auto
      set\ queue\ gpu:gram5.ceres.auckland.ac.nz
      print\ global\ queue
      print\ job\ myjob\ submissionLocation
\f[]
.SS version
.PP
The application package version.
.PP
This is the application package version to be used.
By default the value is `any'.
.PP
Note that this global is not visible under `print globals' unless
it has been set.
To unset the variable use the `unset' command
.PP
If a package is specified and the queue is set to auto, the job
will be submitted to a queue location that supports a version of
the chosen application package.
.PP
To ensure a specific version of the package is used, use the `set'
command to choose the version.
To see the list of versions available for an application package
use the command `print package <application_package>'.
.PP
Example usage:
.PP
\f[CR]
      set\ package\ R
      set\ version\ any
      
      set\ package\ R
      set\ version\ \ 2.11.1
      
      unset\ version
\f[]
.SS walltime
.PP
The walltime for the job measured in minutes.
.PP
The walltime determines the upper limit on how long a job will
execute for.
If a job has not finished after the allocated walltime, the job
will be killed.
.PP
Walltime can be set with strings as follows:
.PP
set walltime 120 : Sets the walltime for 2 hours set walltime
1d2h3m : Sets the walltime for 1 day 2 hours and 3 minutes.
.PP
To view the walltime before a job has been submitted, use the
command `print global walltime'.
To view the walltime after a job has been submitted, use the
command `print job  walltime'.
.PP
Example usage:
.PP
\f[CR]
      set\ walltime\ 240
      set\ walltime\ 240m
      set\ walltime\ 4h
      set\ walltime\ 30d4h12m
      print\ global\ walltime
      print\ job\ myjob\ walltime
\f[]
.SH COMMANDS
.SS about
.PP
Displays the following information about the Gricli shell:
.PP
version: This is the software version you are using.
grisu frontend version: The interface used to communicate with
Grisu.
grisu backend: The Grisu backend (BeSTGRID or DEV) grisu backend
host: The Grisu host.
grisu backend version: The version of Grisu used.
documentation: Where you may find further help and information.
contact: Who to contact in case you have problems or questions.
.PP
Example usage:
.PP
\f[CR]
      about
\f[]
.SS add
.PP
Adds an item to a list.
.PP
Currently only a single item can be added per call.
To add multiple items, use this command once for each item.
.PP
Parameters:
.PP
\f[CR]
      list\ :\ The\ name\ of\ the\ list.
      item\ :\ The\ value\ to\ add.\ 
\f[]
.PP
Currently available lists are:
.PP
\f[CR]
      files\ :\ The\ files\ attached\ for\ a\ job.

CHANGELOG.md
Changelog
=========

0.8
---

* added way to start session daemon in background for Linux (-s option) which keeps credential up-to-date
* 'exit' alias for 'quit' -- both commands can be used to logout now
* improved checks of job properties against valid values -- if a queue is set that doesn't support currently setup job a detailed error message is displayed
* improved information for queues when issuing 'print queue xxxx' command
* 'print queues' now displays queues sorted by whether a queue supports the currently setup job or not
* improved ls command to list jobs easier. format: ls <jobname> <filename_in_jobdirectory>
* changes in speed and stability because of update of underlying Grisu client library


0.7
---

* All gricli properties are now exported as job environment variables, like walltime -> GRISU_WALLTIME
* "kill/clean jobs" commands now deprecated. "kill/clean job [list of jobnames/globs]" is used insted. Supports globs completion.
* "download" and "downloadclean" commands now support optional target directory.
* "view" command can now list jobdirectories directly ("view job <jobname> <file_relative_to_jobdir>").
* Changes to login-component, makes renewal of proxy within gricli possible via 'renew session'.
* Destroy proxy renamed to 'close session'.
* Notification when user proxy reaches configured minimum
* man & markdown USAGE.md pages are created automatically via maven
* Debian and RedHat packages
* "kill/clean", "archive" and "submit" commands can now be executed in background using "&" operator
* Custom job type removed from docs and replaced by single
* hostCount renamed to hostcount
* host is no longer a global
* print queue [propeties] doesn't list memory properties
* added commands to monitor asynchronous tasks

0.6
----

* renamed global "application" to "package"
* fixed job download where directories within job directories weren't downloaded
* updated Grisu dependency which brings lots of improvements to scalability and stability
* "archive job" blocks now until archiving is finished
* progress bar for killing of multiple jobs
* basic "view" command to pre-view files. mostly useful for viewing stdout/stderr while job is running

0.5.2
-----

* updated Grisu dependency with shib login fix

0.5.1
-----

* updated Grisu dependency with more logging statements

0.5
---

* walltime global display has changed to human readable format
* clear files has been replaced by unset files, unset is a new command
* new command: unset - can be used to set optional globals to their null values. unset on required globals will throw error 
* script comments #
* glob support for print package
* new commands clean/kill jobs
* persistent user session; globals except attached files and env vars saved
* last used group remembered
* tab-completion for remote file systems
* tab-completion now also works on "submit" command for executables and input files
* submit command no longer needs quotes
* BeSTGRID-DEV renamed to **dev** case insensitive
* "print queues" now displays only queues that are available for the currently setup environment
* "print queues" can now display more information about queues, like free job slots, total running jobs
* new command: "print queue [queue]" -- displays all information that is known for a queue (which needs to be available for the current environment).
* new command: "status" -- prints out a summary of how many jobs are active/finished/failed
* fixed job status problem where they are incorrectly marked Done.
* renamed application to package
* pwd and exec pwd are the same now
* gls command replaced by ls

0.4
---

* improved internal help system
  - Added commands descriptions and usage examples
  - help can now be requested for the use of a particular command (e.g. help cd)
* overall improvement of ease of use
  - added print groups command
  - added block size support for memory value set (e.g. set memory 3g)
  - added walltime set through formatted time value (e.g. set walltime 1h20m)
* New job type smp added to improve scheduling performance
* Added support for staging in data from remote gridftp servers
   - data can be staged in directly from remote gridftp servers (e.g. to stage in a file from the home directory on data fabric, use command attach grid://Groups/nz/nesi/my_file_here.dat
* jobs can have description property now
* added wait job command to pause gricli and wait until the specified job terminates
* added commands: about, cd, exec, ls, pwd
* improved auto-completion
* patches that fix bugs on Windows

Related Issues
---

Full details of all issues resolved in this release are available here:

July 2011 Milestone: https://github.com/grisu/gricli/issues?state=closed&milestone=1

June 2011 Milestone: https://github.com/grisu/gricli/issues?state=closed&milestone=5


src/main/java/grisu/gricli/Gricli.java
package grisu.gricli;

import ch.qos.logback.classic.LoggerContext;
import ch.qos.logback.classic.joran.JoranConfigurator;
import ch.qos.logback.core.joran.spi.JoranException;
import ch.qos.logback.core.util.StatusPrinter;
import com.beust.jcommander.ParameterException;
import com.google.common.base.Strings;
import grisu.frontend.control.login.LoginManager;
import grisu.frontend.view.cli.GrisuCliClient;
import grisu.gricli.command.GricliCommand;
import grisu.gricli.command.GricliCommandFactory;
import grisu.gricli.command.InteractiveLoginCommand;
import grisu.gricli.command.RunCommand;
import grisu.gricli.completors.CompletionCache;
import grisu.gricli.completors.DummyCompletionCache;
import grisu.gricli.environment.GricliEnvironment;
import grisu.gricli.environment.GricliVar;
import grisu.gricli.parser.GricliTokenizer;
import grisu.jcommons.utils.VariousStringHelpers;
import grisu.jcommons.view.cli.CliHelpers;
import grisu.jcommons.view.cli.LineByLineProgressDisplay;
import grisu.settings.Environment;
import grith.jgrith.control.SlcsLoginWrapper;
import grith.jgrith.cred.GridLoginParameters;
import grith.jgrith.plainProxy.LocalProxy;
import jline.ArgumentCompletor;
import jline.ConsoleReader;
import jline.History;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Date;
import java.util.List;
import java.util.logging.Level;

import static grisu.gricli.GricliExitStatus.*;

public class Gricli extends GrisuCliClient<GricliCliParameters> {

	static final Logger myLogger = LoggerFactory.getLogger(Gricli.class
			.getName());

	static final String CONFIG_FILE_PATH = FilenameUtils.concat(Environment
			.getGrisuClientDirectory().getPath(), "gricli.profile");
	static final String SESSION_SETTINGS_PATH = FilenameUtils.concat(
			Environment.getGrisuClientDirectory().getPath(),
			"gricli-session.profile");

	static final String HISTORY_FILE_PATH = FilenameUtils.concat(Environment
			.getGrisuClientDirectory().getPath(), "gricli.hist");

	static final String DEBUG_FILE_PATH = FilenameUtils.concat(Environment
			.getGrisuClientDirectory().getPath(), "gricli.debug");

	public static final String COMPLETION_CACHE_REGISTRY_KEY = "CompletionCache";

	public static CompletionCache completionCache = new DummyCompletionCache();

	static String scriptName = null;

	static private GricliEnvironment env;
	public static final GricliCommandFactory SINGLETON_COMMANDFACTORY = GricliCommandFactory
			.getCommandFactory();

	static private GricliExitStatus exitStatus = SUCCESS;

	// public static final int MINIMUM_PROXY_LIFETIME_BEFORE_RENEW_REQUEST =
	// (3600 * 24 * 9)
	// + (3600 * 23) + 3500;

	// 3 days
	public static final int MINIMUM_PROXY_LIFETIME_BEFORE_RENEW_REQUEST = 259200;

	private static void configLogging() {
		LoggerContext lc2 = (LoggerContext) LoggerFactory.getILoggerFactory();
		// print logback's internal status
		// StatusPrinter.print(lc2);

		// stop javaxws logging
		java.util.logging.LogManager.getLogManager().reset();
		java.util.logging.Logger.getLogger("root").setLevel(Level.ALL);

		String logback = "/etc/gricli/gricli.log.conf.xml";

		if (!new File(logback).exists() || (new File(logback).length() == 0)) {
			logback = Environment.getGrisuClientDirectory()
					+ File.separator
					+ "gricli.log.conf.xml";
		}
		if (new File(logback).exists()
				&& (new File(logback).length() > 0)) {

			LoggerContext lc = (LoggerContext) LoggerFactory
					.getILoggerFactory();

			try {
				JoranConfigurator configurator = new JoranConfigurator();
				configurator.setContext(lc);
				// the context was probably already configured by default
				// configuration
				// rules
				lc.reset();
				configurator.doConfigure(logback);
			} catch (JoranException je) {
				je.printStackTrace();
			}
			StatusPrinter.printInCaseOfErrorsOrWarnings(lc);

		}


	}

	private static void executionLoop() throws IOException {

		if (System.console() == null) {
			run(System.in);
			return;
		}

		if (scriptName != null) {
			myLogger.debug("Executing script: script=[{}]", scriptName);
			List<String> lines = FileUtils.readLines(new File(scriptName));
			for (Integer i=0; i<lines.size(); i++) {
				String line = lines.get(i);
				myLogger.debug("script=[{}] line=[{}]", line,
						Strings.padStart(i.toString(), 3, '0'));
			}
			run(new FileInputStream(scriptName));
			myLogger.debug("Finished script: script=[{}]", scriptName);
			return;
		}

		final ConsoleReader reader = getReader();
		while (true) {

			String prompt = getPrompt();

			final String line = reader.readLine(prompt);

			if (line == null) {
				break;
			}
			final String[] commandsOnOneLine = line.split(";");
			for (final String c : commandsOnOneLine) {
				runCommand(GricliTokenizer.tokenize(c),
						SINGLETON_COMMANDFACTORY, env);
			}

		}
	}

	private static String generateSession(GricliEnvironment env) {
		String result = "";
		for (final GricliVar<?> var : env.getVariables()) {
			if (var.isPersistent()) {
				final Object value = var.get();
				if (value == null) {
					result += "unset " + var.getName() + "\n";
				} else {
					result += "set " + var.getName() + " "
							+ GricliTokenizer.escape(var.marshall()) + "\n";
				}
			}
		}
		return result;
	}

	private static String getPrompt() {
		String prompt = env.prompt.get();

		// check whether there are new notifications for users.
		if (env.getNotifications().size() > 0) {
			prompt = "(*" + env.getNotifications().size() + ") " + prompt;
		}

		/*
		 * will have to restore this function later for (String var :
		 * env.getGlobalNames()) { prompt = StringUtils.replace(prompt, "${" +
		 * var + "}", env.get(var));
		 *
		 * }
		 */
		return prompt;
	}

	private static ConsoleReader getReader() throws IOException {
		// ConsoleReader reader = new ConsoleReader();
		final ConsoleReader reader = CliHelpers.getConsoleReader();
		reader.setHistory(new History(new File(HISTORY_FILE_PATH)));

		final ArgumentCompletor completor = new ArgumentCompletor(
				SINGLETON_COMMANDFACTORY.createCompletor(),
				new SemicolonDelimiter());
		completor.setStrict(false);
		reader.addCompletor(completor);
		return reader;
	}

	private static boolean login(GricliEnvironment env,
			GrisuCliClient<GricliCliParameters> client) {

		try {
			GricliCommand login = null;
			// if (System.console() != null) {
			login = new InteractiveLoginCommand(client);
			// } else {
			// login = new LocalLoginCommand(client.getCliParameters()
			// .getBackend());
			// }
			login.execute(env);

			prepareLogging(env);

			return true;
		} catch (final Exception ex) {
			myLogger.error("Login exception", ex);
			Throwable t = ex;
			while (t.getCause() != null) {
				t = t.getCause();
			}
			System.err.println(t.getLocalizedMessage());
			return false;
		}
	}


	// private static boolean login(GricliEnvironment env, String backend,
	// String credConfig) {
	//
	// try {
	// GricliCommand login = null;
	// login = new LocalLoginCommand(backend, credConfig, null);
	//
	// login.execute(env);
	//
	// prepareLogging(env);
	//
	// return true;
	// } catch (final Exception ex) {
	// myLogger.error("Login exception", ex);
	// Throwable t = ex;
	// while (t.getCause() != null) {
	// t = t.getCause();
	// }
	// System.err.println(t.getLocalizedMessage());
	// return false;
	// }
	//
	// }

	@SuppressWarnings("static-access")
	public static void main(String[] args) {

		// try {
		// Thread.sleep(3000);
		// } catch (InterruptedException e1) {
		// // TODO Auto-generated catch block
		// e1.printStackTrace();
		// }
		// System.out.println("STARTING...");
		// System.setProperty(
		// CommonGridProperties.Property.DAEMONIZE_GRID_SESSION.toString(),
		// "false");

		Gricli g = null;
		try {
			g = new Gricli(args);
		} catch (Exception e) {
			System.err.println("Could not start gricli: "
					+
					e.getLocalizedMessage());
			System.exit(1);

		}

		g.run();

	}

	private static void prepareLogging(GricliEnvironment env) {
		try {
			final String dn = env.getServiceInterface().getDN();
			MDC.put("dn", dn);
			String un = VariousStringHelpers.getCN(dn);
			MDC.put("user", un);
		} catch (final Exception e) {
			myLogger.error(e.getLocalizedMessage(), e);
		}
	}

	@SuppressWarnings("unchecked")
	private static void run(InputStream in) throws IOException {

		final GricliTokenizer t = new GricliTokenizer(in);
		String[] tokens;
		while ((tokens = t.nextCommand()) != null) {
			runCommand(tokens, SINGLETON_COMMANDFACTORY, env);

		}
	}

	private static void runCommand(String[] c, GricliCommandFactory f,
			GricliEnvironment env) {

		if ((c == null) || (c.length == 0)) {
			return;
		}

		Throwable error = null;
		Date start = null;
		Date end = null;
		try {


			start = new Date();
			String cmdId = c[0] + "_" + start.getTime();
			MDC.put("cmdid", cmdId);

			final GricliCommand command = f.create(c);

			if ((c == null) || (c.length == 0)) {
				return;
			}

			myLogger.info("Executing command: " + "command=[{}]",
					StringUtils.join(c, " "));
			command.execute(env);

			exitStatus = SUCCESS;

		} catch (final InvalidCommandException ex) {
			exitStatus = SYNTAX;
			error = ex;
			System.out.println(ex.getMessage());
		} catch (final UnknownCommandException ex) {
			exitStatus = SYNTAX;
			error = ex;
			System.err.println(ex.getMessage());
		} catch (final SyntaxException ex) {
			exitStatus = SYNTAX;
			error = ex;
			System.err.println("syntax error " + ex.getMessage());
		} catch (final LoginRequiredException ex) {
			exitStatus = LOGIN;
			error = ex;
			System.err.println("this command requires you to login first");
		} catch (final GricliSetValueException ex) {
			exitStatus = RUNTIME;
			error = ex;
			if (StringUtils.isBlank(ex.getValue())) {

				System.err.println("Global '" + ex.getVar()
						+ "' cannot be unset.");
			} else {
				System.err.println("variable " + ex.getVar()
						+ " cannot be set to " + ex.getValue());
				System.err.println("reason: " + ex.getReason());
			}
		} catch (final GricliRuntimeException ex) {
			exitStatus = RUNTIME;
			Throwable exc = ex;
			while (exc.getCause() != null) {
				exc = exc.getCause();
			}
			error = exc;
			System.err.println(exc.getMessage());
		} catch (final RuntimeException ex) {
			exitStatus = RUNTIME;
			error = ex;
			System.err
			.println("command failed. Either connection to server failed, or this is gricli bug. "
					+ "Please send "
					+ DEBUG_FILE_PATH
					+ " to eresearch-admin@auckland.ac.nz together with description of what triggered the problem");
		} finally {

			end = new Date();
			if (start == null) {
				start = end;
			}

			long duration = end.getTime() - start.getTime();

			if (error != null) {
				myLogger.error(error.getLocalizedMessage(), error);
				if (env.debug.get() && (error != null)) {
					error.printStackTrace();
				}
				myLogger.info("Finished command: "
						+ "command=[{}] [failed] error=[{}] duration=[{}]",
						new Object[] { StringUtils.join(c, " "),
								error.getLocalizedMessage(), duration });
			} else {
				myLogger.info("Finished command: "
						+ "command=[{}] duration=[{}]",
						StringUtils.join(c, " "), duration);
			}
			MDC.remove("cmdid");
		}
	}

	public static void shutdown(GricliEnvironment env) {
		try {
			final File f = new File(SESSION_SETTINGS_PATH);
			final String session = generateSession(env);
			FileUtils.writeStringToFile(f, session);
		} catch (final IOException ex) {
			myLogger.error(ex.getLocalizedMessage(), ex);
			env.printError("warning: could not save session");
		}
	}

	private static void startSession() throws IOException {
		try {
			if (new File(SESSION_SETTINGS_PATH).exists()) {
				new RunCommand(SESSION_SETTINGS_PATH).execute(env);
			}
			if (new File(CONFIG_FILE_PATH).exists()) {
				new RunCommand(CONFIG_FILE_PATH).execute(env);
			}
		} catch (final GricliRuntimeException ex) {
			// config does not exist
			env.printError(ex.getMessage());
		}
		executionLoop();
		shutdown(env);
		System.exit(exitStatus.getStatus());
	}

	// private final String[] args;

	public Gricli(String[] args) throws Exception {
		super(new GricliCliParameters(), args);
		// this.args = args;
	}

	@Override
	public void run() {


		configLogging();


		try {
			LoginManager.initGrisuClient("gricli");
		} catch (Exception e) {
			System.err.println("Can't initialize environment: "
					+ e.getLocalizedMessage());
			System.exit(1);
		}


		Thread.setDefaultUncaughtExceptionHandler(new DefaultExceptionHandler());

		CliHelpers.setProgressDisplay(new LineByLineProgressDisplay());



		final Thread t = new Thread() {
			@Override
			public void run() {
				try {
					myLogger.debug("Preloading idps...");
					SlcsLoginWrapper.getAllIdps();
				} catch (final Throwable e) {
					myLogger.error(e.getLocalizedMessage(), e);
				}
			}
		};
		t.setDaemon(true);
		t.setName("preloadIdpsThread");

		if (LocalProxy
				.validGridProxyExists(MINIMUM_PROXY_LIFETIME_BEFORE_RENEW_REQUEST / 60)) {
			// not that important, still, we want it to load in case of renew
			// session
			t.setPriority(Thread.MIN_PRIORITY);
		}
		t.start();

		try {
			GricliCliParameters gcp = getCliParameters();

			env = new GricliEnvironment();
			SigintHandler.install(env);

			GridLoginParameters glp;
			glp = GridLoginParameters.createFromGridCliParameters(gcp);


			if (glp.isNologin()) {
				startSession();
			}


			if (!login(env, this)) {
				System.exit(LOGIN.getStatus());
			}

			scriptName = gcp.getScript();

			startSession();
		} catch (ParameterException pe) {
			// throw new ParseException(pe.getLocalizedMessage());
			pe.printStackTrace();

		} catch (final Throwable th) {
			th.printStackTrace();
			System.err
			.println("Something went terribly wrong.  Please check if you have internet connection, and your firewall settings."
					+ " If you think there is nothing wrong with your connection, send "
					+ DEBUG_FILE_PATH
					+ " to eresearch-admin@auckland.ac.nz together with description of what you are trying to do.");
			myLogger.error("something went terribly wrong ", th);
		}
	}


}

class SemicolonDelimiter extends ArgumentCompletor.AbstractArgumentDelimiter {

	@Override
	public boolean isDelimiterChar(String s, int i) {
		return ((s != null) && (s.length() > i) && (s.charAt(i) == ';'));
	}

}
